### PR TITLE
resize: EVE-patterned resume tests and SIGKILL chaos harness

### PR DESCRIPTION
**Pull request by eriknordmark**
Mostly tests + instrumentation (a build-tagged, no-op-by-default backend hook
used only by the chaos test), plus one block-device production fix in `run.go`:
open the whole disk non-exclusively so child-partition `e2fsck`/`resize2fs` can
run, and accept `e2fsck`'s "errors corrected" exit codes under fix-errors. That
fix is what introduces the go-diskfs dependency above.

## What

Realistic, EVE-shaped resize tests plus a SIGKILL chaos harness that exercises
resume safety under crash injection.

- **EVE-patterned layout tests** (`eve_resume_test.go`): a scaled-down EVE disk
  (FAT32 ESP, squashfs IMGA/IMGB, CONFIG, ext4 P9 at index 9) driven through the
  resume/idempotency harness — a Case-1 in-place P9 shrink and a Case-2 grow of
  ESP/IMGA/IMGB into the freed space — plus byte-integrity, atomic-failure,
  shrink-below-used, and combined shrink+grow invariants, with per-stage GPT
  table logging.
- **SIGKILL chaos** (`eve_stress_test.go`): runs the full two-step EVE resize as
  a subprocess and group-SIGKILLs it at random points across all five pipeline
  steps, then re-runs to completion and asserts the result matches an
  uninterrupted run; records which step each kill hit.
- **GPT-write-delay hook** (`delaybackend_*.go`, `-tags chaos`; `run.go`
  `maybeWrapBackend`): a no-op in normal builds. Under the chaos tag it delays
  around GPT-sector writes so kills can land inside the otherwise-instantaneous
  `updatePartitions`/`createPartitions` table writes.
- **Post-kill GPT integrity** (`gpt_integrity_test.go`): after each kill, checks
  primary/backup header self-CRCs, entry-array CRCs, and primary↔backup entry
  equality, so torn tables and table divergence are observed — and shown to be
  healed by the resume.
- **Copy-dest state** (`copyprogress_test.go`, opt-in): classifies a grow
  target's data region as empty/partial/complete after a kill.

## Results

With the GPT-write delay, kills reach all five pipeline steps including
`updatePartitions`; the resume recovers from every observed torn-entry-array and
primary↔backup-mismatch state.

### CODE REVIEW

**Comment by eriknordmark** [comment]
@deitch ready for review when you have a chance — see the PR description for what it does.

**Comment by eriknordmark** [comment]
@deitch this is ready for review when you have a chance.

To back up the resume-safety claim, I ran an 8-hour SIGKILL chaos soak of `TestEVEChaosKill` with the GPT write-delay hook enabled (`CHAOS_GPT_DELAY=5s`), which widens the window around the otherwise-instantaneous GPT-table writes so kills can land *inside* them. Each iteration runs the full EVE-style two-step resize (shrink P9, then grow ESP/IMGA/IMGB) as a subprocess, SIGKILLs it at a random point, then re-runs to completion and asserts the result is byte-identical to an uninterrupted resize.

**Result: 85 iterations, 0 failures.** Every killed resize resumed cleanly.

Where the 584 kills landed:

| Step | Kills |
|---|---|
| shrinkPartitions | 186 |
| preflight | 167 |
| updatePartitions | 98 |
| createPartitions | 92 |
| copyFilesystems | 28 |
| shrinkFilesystems | 13 |

The delay did its job — 98 kills landed inside `updatePartitions` and 92 inside `createPartitions`, i.e. mid-GPT-write.

On-disk GPT integrity sampled at each kill (584 samples):

| State | Count |
|---|---|
| Header self-CRC failure (primary or backup) | **0** |
| Torn primary entry array (CRC mismatch) | 66 |
| Torn backup entry array (CRC mismatch) | 64 |
| Primary↔backup entry mismatch | 121 |
| Clean | 395 |

So both tear shapes occur in practice — a torn 32-sector entry array, and a primary/backup ordering skew (backup written first, primary not yet) — and the resume logic heals every one. Header self-CRCs never tear because the header is a single 512-byte sector and its write is atomic, confirmed across all 584 samples.

File: README.md [hunk 1]
```diff
@@ -52,3 +52,37 @@ Grow partition named sda2 to 50G on disk image file disk.img:
 resizer --grow-partition name:sda2:50G disk.img
 ```
 
+## Testing
+
+Most tests run with `go test ./...`. Several end-to-end tests perform real
+shrinks/copies of multi-GB fixtures and are slow; skip them in a fast loop with
+`go test -short ./...`.
+
```
Comment (deitch) [inline README.md:153]: Can we put a brief list of all of the test options here? As far as I can tell, right now there are:

* `go test ./...` - includes all of the normal tests, including the shrinks/copies which can take up disk space and time
* `go test -short ./...` - same as above, but skips the shrinks/copies
* `RESIZER_CHAOS=1` - enables a set of chaos tests, on top of the usual ones
* `CHAOS_SEED` - normally random, but allows you to reproduce if needed (I rather like this idea)

Anything else? 

I also find it a bit confusing, that the default is lots of big tests, but not *all* of them, but go minimum with `-short`, but then enable more with `RESIZER_CHAOS`. I know that `-short` is a normal way to do it, and then using env vars is normal for adding more (other than tags, which might or might not be any better, e.g. `go test -tags=chaos`). But it still is a little confusing.

File: README.md [hunk 2]
```diff
@@ -52,3 +52,37 @@ Grow partition named sda2 to 50G on disk image file disk.img:
...
+### Chaos / resume soak test
+
+`TestEVEChaosKill` is a stress test, not a CI gate. It repeatedly runs the full
```
Comment (deitch) [inline README.md]: I am not particularly enamoured of calling these "EVE tests". I get that you are testing a layout similar (identical?) to what EVE has. But partitionresizer is no more EVE-specific than diskfs is linuxkit-specific. It was the original driver for the idea, but built to be more generic. Even if it is identical, calling it something else that addresses the general flow would be better.

File: copyprogress_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	"io"
+	"os"
+	"sort"
+	"strings"
+)
+
+// gptPartitions parses the primary GPT entry array (LBA 1 header) and returns a
+// map of partition label -> [firstLBA, lastLBA]. Used to locate a grow target
+// and its source on disk without depending on the in-memory fixture layout.
+func gptPartitions(f *os.File) map[string][2]int64 {
```
Comment (deitch) [inline copyprogress_test.go]: Here (and in a few other places), why would we recreate all of the work of reading the GPT info? Why not just use the capabilities of go-diskfs?

**Review by deitch — COMMENTED** [review]
Some questions.

File: README.md (hunk 1 above)
Comment (eriknordmark) [reply README.md:153]: Do you just want a table documenting it, or change the chaos to go test -chaos ./... with the chaos-specific envs changed into command line argument?

Isn't -tags=chaos more of a build thing than a test running thing?

Comment (deitch) [reply README.md:153]: Yeah `-tags=chaos` often is used for builds rather than tests, but I have seen for both. As I said above, there is no great answer. Maybe just leave it like it is, and we can kvetch about it later.

Comment (eriknordmark) [reply README.md:153]: Let me show you an option with -chaos and if you don't like that one I'll just add documentation of what we have.

Comment (eriknordmark) [reply README.md:153]: Added a "Testing" section with two tables — the run tiers (`go test ./...` / `-short` / chaos soak) and the chaos env knobs (`RESIZER_CHAOS`, `CHAOS_SEED`, `CHAOS_GPT_DELAY`, `CHAOS_COPY_STATE`).

One correction on the layering: `-tags chaos` isn't something you pass to `go test` — the chaos test builds the resizer *subprocess* with it internally when `CHAOS_GPT_DELAY` is set, so the only user-facing knobs are `-short` plus those env vars.

On "a bit confusing": agreed, and I think a possible fix is to match the `-short` idiom you point at. I'll convert the env vars to test flags — `-chaos` to run the soak, `-chaos.gpt-delay=5s`, `-chaos.seed=N`, `-chaos.copy-state` — and register a no-op `-chaos` in `cmd/resizer` too, since custom flags (unlike the built-in `-short`) only exist in the package that declares them, so `go test -chaos ./...` would otherwise error there. README will track whichever form we land on.

File: README.md (hunk 2 above)
Comment (eriknordmark) [reply README.md]: Agreed, the name overfits. What's actually under test is a representative mixed-filesystem layout — FAT32 ESP, raw squashfs read-only images, ext4 data — exercised through a shrink-the-tail / grow-the-head flow. EVE was just the layout that motivated it. I'll rename the files and symbols (`eve_*` → layout/resume, `TestEVE*` → `TestMixedLayout*` / `TestChaos*`, `buildEVEFixture` → `buildMixedLayoutFixture`) and reword the README so EVE appears only as the motivating example.

File: copyprogress_test.go (hunk 3 above)
Comment (eriknordmark) [reply copyprogress_test.go]: These run on a post-SIGKILL disk that may have a torn primary, a torn backup, or a primary↔backup divergence. go-diskfs's `gpt.Read` validates CRCs and silently falls back to the backup when the primary fails (`RecoveredFromBackup`), so going through it would heal away exactly the torn/divergent states these tests exist to observe. Where the disk is known-good we already use go-diskfs (`gptByName` / `gptDump` and the stress harness go through `GetPartitionTable`). I'll consolidate the raw readers into one tolerant helper with a comment explaining why this path deliberately can't use go-diskfs.

File: README.md (hunk 1 above)
Comment (eriknordmark) [reply README.md:153]: @deitch let me know if you want the -chaos variants, or whether documenting this is sufficient.

**Comment by eriknordmark** [comment]
I cleanup up the commits including removing intermediate references to old names (of scripts etc). No content changes in HEAD.

File: copyprogress_test.go (hunk 3 above)
Comment (deitch) [reply copyprogress_test.go]: That works, thanks

**Comment by deitch** [comment]
How does this look? Ready to go?

**Review by deitch — APPROVED** [review]

**Comment by eriknordmark** [comment]
I'm doing integration testing with EVE and that found an issue or two in go-diskfs fixed in #411, which is now merged. partitionresizer is re-pinned onto it in #20; once that lands we can run this PR through CI.

**Comment by deitch** [comment]
I guess bring in 411 now?

**Comment by eriknordmark** [comment]
Done — the re-pin is up as #20.
